File: ttextract.py
import os, sys
from unlz2k import unlz2k
unpack = True

# Via http://www.isthe.com/chongo/tech/comp/fnv/
FNV_BASIS = 2166136261
#FNV_PRIME = 16777619
FNV_PRIME = 1677619

algs = {0: '----', 2: 'LZ2K'}

def get_extract_name(file):
    extract_name = os.path.basename(file.name)
    ext = extract_name.rfind('.')
    return extract_name[:ext]

def read_name(file, offset):
    """ Reads a string from the file at the offset (null terminated) """
    prev_offset = file.tell()
    file.seek(offset)
    item_name = ""
    char = int.from_bytes(file.read(1), "little")
    while char != 0:
        item_name += chr(char)
        char = int.from_bytes(file.read(1), "little")
    file.seek(prev_offset)
    return item_name

def handle_fpk(file, file_size):
    # Number of files to read
    num_files = int.from_bytes(file.read(4), byteorder="little")

    # Total file size
    expected_size = int.from_bytes(file.read(4), byteorder="little")
    if file_size != expected_size:
        print('Error: File size mismatch (expected {:<8X}, got {:<8X})'.format(expected_size, file_size))
        exit(0)
    for i in range(num_files):
        name_offset = int.from_bytes(file.read(4), byteorder="little")
        data_offset = int.from_bytes(file.read(4), byteorder="little")
        file_size = int.from_bytes(file.read(4), byteorder="little")
        sixteen = int.from_bytes(file.read(4), byteorder="little")
        if sixteen != 16:
            print('Error: Expected 16 but got {}'.format(sixteen))
            exit(0)
        # Read three words of nothing
        file.read(12)
        name = read_name(file, name_offset)
        # Split everything by '\' and extract directory
        directories = name.rfind('\\')
        file_dir = name[0:directories]
        # Create file
        print('{:<8X}\t{:<8X}\t{}\t{}'.format(data_offset, file_size, i, name))
        dirs_exist = os.path.exists(file_dir)
        if not dirs_exist:
            os.makedirs(file_dir)
        new_file = open(name, 'wb')
        # Jump to data offset
        prev_offset = file.tell()
        file.seek(data_offset)
        new_file.write(file.read(file_size))
        file.seek(prev_offset)
    return

def handle_dat(file, file_size, first_word):

    # Get basename
    extract_name = get_extract_name(file)

    # File info section
    
    file_info_offset = first_word
    file_info_size = int.from_bytes(file.read(4), byteorder="little")
    expected_size = file_info_offset + file_info_size
    if expected_size != file_size:
        print('Error: File size mismatch (expected {:<8X}, got {:<8X})'.format(expected_size, file_size))
        exit(1)
    file.seek(file_info_offset)
    signature = int.from_bytes(file.read(4), byteorder="little", signed=True)
    if signature not in [-1, -3]:
        print('Error: File info signature invalid.')
        exit(1)
    print('DAT file with signature: {}'.format(signature))
    num_files = int.from_bytes(file.read(4), byteorder="little")
    file_info_offset += 8
    print('File info offset: {:<8X}'.format(file_info_offset))
    print('File info size: {:<8X}'.format(file_info_size))
    print('Number of files: {}'.format(num_files))

    # Name info section

    name_info_offset = file_info_offset + (num_files * 16)
    file.seek(name_info_offset)
    num_names = int.from_bytes(file.read(4), byteorder="little")
    name_info_offset += 4
    print('Name info offset: {:<8X}'.format(name_info_offset))
    print('Number of names: {}'.format(num_names))

    # Name data section

    name_data_offset = name_info_offset + (num_names * 8)
    file.seek(name_data_offset)
    # 4 bytes preceding name data is the CRC offset.
    name_crc_offset = int.from_bytes(file.read(4), byteorder="little")
    name_data_offset += 4
    name_crc_offset += name_data_offset
    print('Name data offset: {:<8X}'.format(name_data_offset))
    print('Name CRC offset: {:<8X}'.format(name_crc_offset))

    # CRC section
    
    file_index_of_crc = dict() # Key is CRC, value is file index
    has_crcs = True
    file.seek(name_crc_offset)
    first = int.from_bytes(file.read(4), byteorder="little")
    if first == 0:
        # No CRCs
        has_crcs = False
    if has_crcs:
        file_index_of_crc[first] = 0
        for i in range(1, num_files):
            crc = int.from_bytes(file.read(4), byteorder="little")
            file_index_of_crc[crc] = i
    # Should have two words left, expecting zeroes
    temp = int.from_bytes(file.read(8), byteorder="little")
    if temp != 0:
        print("Error: File bytes at end were unexpected. ({:<8X})".format(file.tell() - 8))
        exit(1)
    extra_data = file.read()
    if extra_data != b'':
        print("Error: Unexpected end of file. ({:<8X})".format(file.tell()))
        exit(1)
    print('Number of CRCs: {}\n'.format(len(file_index_of_crc)))
    print('Offset  \tPacked  \tUnpacked\tAlg?\tFile')
    print('-' * 100)

    """ Now the extraction section!

    Name info table items are 8 bytes long.

    First value is signed 16 bit.
        If positive, read folder name. Value is index for last item in folder.
        If negative or zero, read file name. Value is 2s complement of file ID.

    Second value is signed 16 bit.
        If positive, use previous file path. Value is index of previous item written in directory
        If zero, do not use previous file path. Write item to current directory.

    Third value is relative 32-bit offset to read name from. This is added to the name data offset.
    """

    file.seek(name_info_offset)
    data_offsets = dict()
    file_sizes = dict()
    file_sizes_unpacked = dict()
    last_item_offset = 0
    last_file_id_written = 0 # Unused thus far?
    current_dir = "" # Change this to modify output dir
    item_dirs = dict()
    for name_offset in range(num_names):
        is_folder = False
        is_packed = False
        item_dir = ""
        file_id = -1
        read_type = int.from_bytes(file.read(2), byteorder="little", signed=True)
        if read_type > 0:
            last_item_offset = read_type
            is_folder = True
        else:
            file_id = read_type * -1
        path_type = int.from_bytes(file.read(2), byteorder="little", signed=True)
        if path_type > 0:
            item_dir = item_dirs.get(path_type)
        else:
            item_dir = current_dir
        item_dirs[name_offset] = item_dir
        name_data_relative_offset = int.from_bytes(f.read(4), byteorder="little")

        # Get name
        item_name = read_name(file, name_data_offset + name_data_relative_offset)
        # Append to file path
        if item_name != "": # Avoid creating double slash for blank directory
            item_dir += '\\'
        item_dir += item_name

        if is_folder:
            if item_dir != "":
                # Switch to new directory if it's a folder name
                current_dir = item_dir
        else:
            if has_crcs:
                analysis_name = item_dir[1:].upper()
                crc = FNV_BASIS
                for char in analysis_name:
                    crc = (crc ^ ord(char)) * FNV_PRIME & 0xFFFFFFFF
                # Now find the corresponding CRC index in our collection
                file_index = file_index_of_crc.get(crc)
                if file_index is None:
                    print('Error: File does not appear to have a CRC in the table.')
                    exit(1)
            else:
                # Extract current file
                file_index = file_id

            """ File info table items are 16 bytes long.

            4 bytes = absolute offset of data
                Signature -3 multiplies it by 256
            4 bytes = size of file (archived)
            4 bytes = size of file (uncompressed)
                Signature -1 expects this to match
            3 bytes = packed type
                Signature -1 expects 0
                Signature -3 expects 2 thus far
            1 byte  = fine offset from the given offset
                Signature -1 expects this to be 0
            """
            file_index = file_info_offset + file_index * 16
            prev_offset = file.tell()
            file.seek(file_index)
            if signature == -1:
                offset = int.from_bytes(file.read(4), byteorder="little")
                size_packed = int.from_bytes(file.read(4), byteorder="little")
                size_unpacked = int.from_bytes(file.read(4), byteorder="little")
                if (size_packed != size_unpacked):
                    print('Error: Unknown situation, file sizes do not match. (file {})'.format(i))
                    exit(1)
                packed_type = int.from_bytes(file.read(3), byteorder="little")
                if packed_type != 0:
                    print('Error: Unknown situation, packed type not 0. (file {})'.format(i))
                    exit(1)
                offset += int.from_bytes(file.read(1), byteorder="little")
                data_offsets[file_index] = offset
                file_sizes[file_index] = size_packed
            else:
                offset = int.from_bytes(file.read(4), byteorder="little") << 8
                size_packed = int.from_bytes(file.read(4), byteorder="little")
                size_unpacked = int.from_bytes(file.read(4), byteorder="little")
                packed_type = int.from_bytes(file.read(3), byteorder="little")
                offset += int.from_bytes(file.read(1), byteorder="little")
                if (size_packed != size_unpacked):
                    is_packed = True
                data_offsets[file_index] = offset
                # Right now we're writing compressed files because we don't have the alg
                file_sizes[file_index] = size_packed
                file_sizes_unpacked[file_index] = size_unpacked
            output_dir = extract_name + current_dir
            output_item = extract_name + item_dir
            print('{:<8X}\t{:<8X}\t{:<8X}\t{}\t{}'.format(offset, size_packed, size_unpacked, algs[packed_type], output_item))
            dirs_exist = os.path.exists(output_dir)
            if not dirs_exist:
                os.makedirs(output_dir)
            new_file = open(output_item, 'wb')
            #data_offset = data_offsets.get(file_index)
            # Jump to data offset
            #file.seek(data_offset)
            file.seek(offset)
            if unpack and is_packed:
                if packed_type == 2:
                    unlz2k(file, new_file, size_packed, size_unpacked)
                else:
                    print('Unknown packed type, unpacking raw file')
                    new_file.write(file.read(size_packed))
            else:
                new_file.write(file.read(size_packed))
            # Return to original offset
            file.seek(prev_offset)
            
def handle_default():
    print('Error: Not a recognized .DAT or .FPK file.')
    exit(1)

# First read file
file_name = str(sys.argv[1])
if '-r' in sys.argv:
    unpack = False
try:
    f = open(file_name, "rb")
except:
    print('Error opening the specified file.')
    exit(1)

f.seek(0, 2)
file_size = f.tell()
f.seek(0)

# Read first word. It will tell us what kind of file it is
first_word = int.from_bytes(f.read(4), byteorder="little")
if first_word == 0x12345678:
    handle_fpk(f, file_size)
elif first_word < file_size:
    handle_dat(f, file_size, first_word)
else:
    handle_default()


















    

File: ttextract/ttextract.cpp
// ttextract.cpp : This file contains the 'main' function. Program execution
// begins and ends there.
//

#include "ttextract.h"
#include "include\filereading.h"
#include "include\unlz2k.h"
#include <cctype>
#include <filesystem>
#include <format>
#include <iostream>
#include <string>
#include <unordered_map>
#include <vector>

bool initialized = false;
char *outfileBuffer;

// Via http://www.isthe.com/chongo/tech/comp/fnv/
constexpr unsigned int FNV_BASIS = 2166136261;
// constexpr unsigned int FNV_PRIME = 16777619;
constexpr unsigned int FNV_PRIME = 1677619;

constexpr std::string nameOfAlg(int packedType) {
  switch (packedType) {
    using enum algType;
  case 0:
    return "----";
  case 2:
    return "LZ2K";
  default:
    return "????";
  }
}

const std::vector<int> validSignatures{-1, -2, -3, -4};

void init() {
  // 1 MB file output buffer
  outfileBuffer = new char[0x100000];
}

int main(int argc, char *argv[]) {
  if (!initialized) {
    init();
    initialized = true;
  }
  cmdlineArgs args;
  try {
    args = parseArgs(argc, argv);
  } catch (int errorCode) {
    return errorCode;
  }
  std::ifstream in(static_cast<std::string>(args.fileName),
                   std::ios::in | std::ios::binary);
  if (!in) {
    std::cerr << "Cannot open source file.\n";
    return 1;
  }
  if (args.isUnpack) {
    // TODO: Replace
    std::ofstream dest(args.outName, std::ios::out | std::ios::binary);
    unlz2k(in, dest);
    return 0;
  }
  in.seekg(0, std::ios_base::end);
  size_t fileSize = in.tellg();
  in.seekg(0);
  // First word of file should determine what type of file we're extracting
  int firstWord = readUint32(in, ENDIAN::little);
  if (firstWord == 0x12345678) {
    // Assume .FPK
    try {
      handleFPK(in, fileSize, args);
    } catch (int errorCode) {
      std::cerr << "Program exited with code " << errorCode << '\n';
      return errorCode;
    } catch (const std::exception &e) {
      std::cerr << "Error: " << e.what();
      return 1;
    }
    return 0;
  } else {
    // Assume .DAT
    uint32_t fileInfoOffset = firstWord;
    // If larger than end of file, try 2s complement >> 8
    if (firstWord > fileSize) {
      fileInfoOffset = (~fileInfoOffset + 1) << 8;
    }
    try {
      handleDAT(in, fileSize, args, fileInfoOffset);
    } catch (int errorCode) {
      std::cerr << "Program exited with code " << errorCode << '\n';
      return errorCode;
    } catch (const std::exception &e) {
      std::cerr << "Error: " << e.what();
      return 1;
    }
    return 0;
  }
  logError("Unrecognized archive type.");
  return 1;
}

std::string readName(std::ifstream &src, size_t offsetInFile) {
  auto currOffset = src.tellg();
  src.seekg(offsetInFile);
  std::string itemName = "";
  char buffer;
  src.read(&buffer, 1);
  while (buffer) {
    itemName += buffer;
    src.read(&buffer, 1);
  }
  src.seekg(currOffset);
  return itemName;
}

void handleFPK(std::ifstream &src, size_t fileSize, cmdlineArgs &args) {
  return;
}

void handleDAT(std::ifstream &src, size_t fileSize, cmdlineArgs &args,
               unsigned int firstWord) {

  // First, the file info section
  auto fileInfoOffset = firstWord;
  auto fileInfoSize = readUint32(src, ENDIAN::little);
  auto expectedSize = fileInfoOffset + fileInfoSize;
  if (expectedSize != fileSize) {
    std::cerr << std::format("Size mismatch. Expected 0x{:<8X}, got 0x{:<8X}",
                             expectedSize, fileSize);
    throw 1;
  }
  src.seekg(fileInfoOffset);
  auto signature = readInt32(src, ENDIAN::little);
  if (!std::count(validSignatures.begin(), validSignatures.end(), signature)) {
    std::cerr << std::format("File signature {} invalid.\n", signature);
    throw 1;
  }
  std::cout << "DAT file with signature: " << signature << '\n';
  auto numFiles = readUint32(src, ENDIAN::little);
  fileInfoOffset += 8;
  std::cout << std::format("File info offset: 0x{:<8X}\n", fileInfoOffset);
  std::cout << std::format("File info size: 0x{:<8X}\n", fileInfoSize);
  std::cout << std::format("Number of files: {}\n", numFiles);

  // Name info section
  auto nameInfoOffset = fileInfoOffset + (numFiles << 4);
  src.seekg(nameInfoOffset);
  auto numNames = readUint32(src, ENDIAN::little);
  nameInfoOffset += 4;
  std::cout << std::format("Name info offset: 0x{:<8X}\n", nameInfoOffset);
  std::cout << std::format("Number of names: {}\n", numNames);

  // Name data section
  auto nameDataOffset = nameInfoOffset + (numNames << 3);
  src.seekg(nameDataOffset);
  auto nameCRCOffset = readUint32(src, ENDIAN::little);
  nameDataOffset += 4;
  nameCRCOffset += nameDataOffset;
  std::cout << std::format("Name data offset: 0x{:<8X}\n", nameDataOffset);
  std::cout << std::format("Name CRC offset: 0x{:<8X}\n", nameCRCOffset);

  // CRC section
  std::unordered_map<unsigned int, unsigned short int> crcToIndex;
  bool hasCRCs = false;
  src.seekg(nameCRCOffset);
  if (src.tellg() != fileSize) {
    auto first = readUint32(src, ENDIAN::little);
    if (first) {
      hasCRCs = true;
      crcToIndex[first] = 0;
      for (uint32_t i = 1; i < numFiles; ++i) {
        auto crc = readUint32(src, ENDIAN::little);
        crcToIndex[crc] = i;
      }
      // Should have two dwords left, expecting zeroes
      unsigned int end = readUint32(src, ENDIAN::little);
      end += readUint32(src, ENDIAN::little);
      if (end) {
        logError(std::format("Unexpected non-zero bytes at 0x{:<8X}\n",
                             static_cast<int>(src.tellg()) - 8));
        throw 1;
      }
      if (src.tellg() != fileSize) {
        logError(std::format("Unexpected non-zero data at 0x{:<8X}\n",
                             static_cast<int>(src.tellg())));
        throw 1;
      }
    }
  }
  std::cout << std::format("Number of CRCs: {}\n\n", crcToIndex.size());
  std::cout << "Offset  \tPacked  \tUnpacked\tAlg?\tFile\n";
  std::cout << std::string(100, '-') << '\n';
  /*
  Now the extraction section!

  Name info table items are 8 bytes long.

  First value is signed 16 bit.
      If positive, read folder name. Value is index for last item in folder.
      If negative or zero, read file name. Value is 2s complement of file ID.

  Second value is signed 16 bit.
      If positive, use previous file path. Value is index of previous item
  written in directory If zero, do not use previous file path. Write item to
  current directory.

  Third value is relative 32-bit offset to read name from. This is added to the
  name data offset.
  */

  src.seekg(nameInfoOffset);
  uint16_t lastItemOffset = 0;
  std::string currentDir{""};
  std::unordered_map<uint16_t, std::string> itemDirs;
  for (uint16_t nameOffset = 0; nameOffset < numNames; ++nameOffset) {
    bool isFolder = false, isPacked = false;
    std::string itemName{""};
    int fileID = -1;
    int16_t readType = readInt16(src, ENDIAN::little);
    if (readType > 0) {
      lastItemOffset = readType;
      isFolder = true;
    } else {
      fileID = readType * -1;
    }
    int16_t pathType = readInt16(src, ENDIAN::little);
    if (pathType > 0) {
      itemName = itemDirs.at(pathType);
    } else {
      itemName = currentDir;
    }
    itemDirs[nameOffset] = itemName;
    uint32_t nameDataRelativeOffset = readUint32(src, ENDIAN::little);

    // Get name
    auto name = readName(src, static_cast<size_t>(nameDataOffset) +
                                      nameDataRelativeOffset);
    if (!name.empty()) {
      itemName += '\\';
    }
    itemName += name;

    if (isFolder) {
      if (!itemName.empty()) {
        currentDir = itemName;
      }
    } else {
      uint32_t fileIndex;
      if (hasCRCs) {
        auto analysisName = itemName.substr(1);
        uint32_t crc = FNV_BASIS;
        for (auto &c : analysisName) {
          crc = (crc ^ toupper(c)) * FNV_PRIME & 0xFFFFFFFF;
        }
        try {
          fileIndex = crcToIndex.at(crc);
        } catch (std::out_of_range error) {
          logError(
              std::format("CRC 0x{:<8X} doesn't correspond to a file", crc));
          throw 1;
        }
      } else {
        // Extract current file
        fileIndex = fileID;
      }
      /*
      File info table items are 16 bytes long.

      4 bytes = absolute offset of data
          Signature other than -1 shifts left 8 bits
      4 bytes = size of file (archived)
      4 bytes = size of file (uncompressed)
      1 byte  = packed type
      2 bytes = unknown??
      1 byte  = byte offset from the given offset
      */
      uint32_t offset;
      uint32_t packedSize;
      uint32_t unpackedSize;
      uint32_t packedType;
      fileIndex = fileInfoOffset + (fileIndex << 4);
      auto prevOffset = src.tellg();
      src.seekg(fileIndex);
      offset = readUint32(src, ENDIAN::little);
      if (signature != -1) {
        offset <<= 8;
      }
      packedSize = readUint32(src, ENDIAN::little);
      unpackedSize = readUint32(src, ENDIAN::little);
      auto nextFour = readUint32(src, ENDIAN::little);
      packedType = nextFour & 0xFF;
      offset += nextFour >> 24;
      isPacked = packedSize != unpackedSize;
      std::string outputDir = args.outDir + currentDir;
      std::string outputItem = args.outDir + itemName;
      std::cout << std::format("{:0>8X}\t{:<8X}\t{:<8X}\t{}\t{}\n", offset,
                               packedSize, unpackedSize, nameOfAlg(packedType),
                               outputItem);
      if (!std::filesystem::is_directory(outputDir)) {
        std::filesystem::create_directories(outputDir);
      }
      std::ofstream dest(outputItem, std::ios::out | std::ios::binary);
      if (!dest) {
        logError("Error opening destination file.");
        throw 1;
      }
      src.seekg(offset);
      if (!args.isRaw && isPacked) {
        if (packedType == 2) {
          unlz2k(src, dest, packedSize, unpackedSize);
        } else {
          logWarning(std::format("Unknown packed type {}, unpacking raw file",
                                 packedType));
          writeToDest(src, dest, packedSize);
        }
      } else {
        writeToDest(src, dest, packedSize);
      }
      // Return to previous offset
      src.seekg(prevOffset);
    }
  }
}

void writeToDest(std::ifstream &src, std::ofstream &dest, size_t numBytes) {
  while (numBytes > 0) {
    size_t toWrite = numBytes < 0x100000 ? numBytes : 0x100000;
    src.read(outfileBuffer, toWrite);
    dest.write(outfileBuffer, toWrite);
    numBytes -= toWrite;
  }
}